### PR TITLE
Webauthn sample fixes

### PR DESCRIPTION
**Pull request by adam-fowler**
- Changes required since underlying webauthn library changed APIs
- Removed handlers so I could store webauthn manager in controller (discovered limitations of handlers)
- Added separate authenticator for user
- Base64 encode challenge in session object. Added to and from function to convert between session object and authentication state

### CODE REVIEW

File: webauthn/Sources/App/Controllers/WebAuthnController.swift [hunk 1]
```diff
@@ -20,167 +20,126 @@ import HummingbirdCore
 import WebAuthn
 
 struct HBWebAuthnController {
+    let webauthn: WebAuthnManager
+
     func add(_ group: HBRouterGroup) {
```
Comment (Joannis) [inline webauthn/Sources/App/Controllers/WebAuthnController.swift:25]: I know this isn't in scope for this PR. But I find this function extremely confusing to read, especially in `Application+configure.swift`. Can we add a label `to`, so it becomes:

```swift
func add(to group: HBRouterGroup)
```

.. and on the call site:

```swift
controller.add(to: self.router.group("api"))
```

File: webauthn/Sources/App/Controllers/WebAuthnController.swift [hunk 2]
```diff
@@ -20,167 +20,126 @@ import HummingbirdCore
...
-
-        typealias Output = HBResponse
-
-        let input: Input
-
-        init(from request: HBRequest) throws {
-            self.input = try request.decode(as: Input.self)
-        }
+    struct SignInInput: Decodable {
+        let name: String
```
Comment (Joannis) [inline webauthn/Sources/App/Controllers/WebAuthnController.swift:40]:
```suggestion
        let username: String
```

File: webauthn/Sources/App/Controllers/WebAuthnController.swift [hunk 3]
```diff
@@ -20,167 +20,126 @@ import HummingbirdCore
...
+        let input = try request.decode(as: SignInInput.self)
+        guard try await User.query(on: request.db)
+            .filter(\.$username == input.name)
+            .first() == nil
+        else {
+            throw HBHTTPError(.conflict, message: "Username already taken.")
         }
+        let user = User(username: input.name)
+        try await user.save(on: request.db)
+        let session = WebAuthnSessionStateAuthenticator.Session.signedUp(userId: user.id!)
```
Comment (Joannis) [inline webauthn/Sources/App/Controllers/WebAuthnController.swift:53]: I know you just saved the user, but especially in an example it's not good to force unwrap variables. Users _may_ define their IDs as user generated, and run into a confusing crash that way. Use `.requireId()` instead, which I believe is a FluentKit API

File: webauthn/Sources/App/Controllers/WebAuthnController.swift [hunk 4]
```diff
@@ -20,167 +20,126 @@ import HummingbirdCore
...
+        try await request.session.save(
+            session: session,
+            expiresIn: .minutes(10)
+        )
+        return .redirect(to: "/api/beginregister", type: .temporary)
```
Comment (Joannis) [inline webauthn/Sources/App/Controllers/WebAuthnController.swift:58]: `/api/register/start` would be a more logical path to me

File: webauthn/Sources/App/Controllers/WebAuthnController.swift [hunk 5]
```diff
@@ -20,167 +20,126 @@ import HummingbirdCore
...
+    let webauthn: WebAuthnManager
+
     func add(_ group: HBRouterGroup) {
         group
-            .post("signup", options: .editResponse, use: SignupHandler.self)
+            .post("signup", options: .editResponse, use: self.signin)
             .get("login", options: .editResponse, use: self.beginAuthentication)
+        group
+            .add(middleware: WebAuthnSessionStateAuthenticator())
+            .post("beginregister", use: self.beginRegistration)
```
Comment (Joannis) [inline webauthn/Sources/App/Controllers/WebAuthnController.swift:31]: `/register/start`

File: webauthn/Sources/App/Controllers/WebAuthnController.swift [hunk 6]
```diff
@@ -20,167 +20,126 @@ import HummingbirdCore
...
+            .post("finishregister", use: self.finishRegistration)
```
Comment (Joannis) [inline webauthn/Sources/App/Controllers/WebAuthnController.swift:32]: `/register/finish`

File: webauthn/Sources/App/Controllers/WebAuthnController.swift [hunk 7]
```diff
@@ -20,167 +20,126 @@ import HummingbirdCore
...
-            let session = WebAuthnSessionAuthenticator.Session.authenticated(userId: webAuthnCredential.user.id!)
-            try await request.session.save(session: session, expiresIn: .hours(24))
-
-            return .ok
+        request.logger.info("Challenge: \(challenge)")
+        do {
+            _ = try self.webauthn.finishAuthentication(
+                credential: input,
+                expectedChallenge: challenge,
+                credentialPublicKey: [UInt8](webAuthnCredential.publicKey.decoded!),
```
Comment (Joannis) [inline webauthn/Sources/App/Controllers/WebAuthnController.swift:126]: force unwrap, can we get rid of it?

File: webauthn/Sources/App/Middleware/WebAuthnAuthenticator.swift [hunk 8]
```diff
@@ -16,34 +16,67 @@ import Foundation
...
+
+        /// init session object from authentication state
+        init(from session: AuthenticationSession) {
+            switch session {
+            case .authenticating(let challenge):
+                self = .authenticating(encodedChallenge: challenge.base64URLEncodedString().asString())
+            case .signedUp(let user):
+                self = .signedUp(userId: user.id!)
+            case .registering(let user, let challenge):
+                self = .registering(userId: user.id!, encodedChallenge: challenge.base64URLEncodedString().asString())
```
Comment (Joannis) [inline webauthn/Sources/App/Middleware/WebAuthnAuthenticator.swift:44]: `user.requireId()`

**Review by Joannis — CHANGES_REQUESTED** [review]

File: webauthn/Sources/App/Controllers/WebAuthnController.swift (hunk 1 above)
Comment (adam-fowler) [reply webauthn/Sources/App/Controllers/WebAuthnController.swift]: done

File: webauthn/Sources/App/Controllers/WebAuthnController.swift (hunk 7 above)
Comment (adam-fowler) [reply webauthn/Sources/App/Controllers/WebAuthnController.swift]: If it fails to decode then we saved bad data into the database, I've changed it to return a server error.

File: webauthn/Sources/App/Controllers/WebAuthnController.swift (hunk 2 above)
Comment (adam-fowler) [reply webauthn/Sources/App/Controllers/WebAuthnController.swift]: username, name I keep swapping between the two. I've changed this one as well

**Review by Joannis — APPROVED** [review]

**Pull request merged**